Add system variable tidb_tso_client_rpc_mode (#19057)
pingcap/docs: system-variables.md

> **Note:**
>
> - Suppose that the TSO RPC latency increases for reasons other than a CPU usage bottleneck of the PD leader (such as network issues). In this case, enabling the TSO Follower Proxy might increase the execution latency in TiDB and affect the QPS performance of the cluster.

> > - This feature is incompatble with [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840). If this feature is enabled, [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840) won't take effect.

Reviewer: Use present tense for statements that describe general behavior. 

https://developers.google.cn/style/tense

> **Note:**
>
> - Suppose that the TSO RPC latency increases for reasons other than a CPU usage bottleneck of the PD leader (such as network issues). In this case, increasing the value of `tidb_tso_client_batch_max_wait_time` might increase the execution latency in TiDB and affect the QPS performance of the cluster.
> - This feature is incompatble with [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840). If this variable is set to non-zero value, [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840) won't take effect.

> ### `tidb_tso_client_rpc_mode` <span class="version-mark">New in v8.4.0</span>

Reviewer: In this English variable doc, the variable in a section title is not wrapped by ` to avoid the search not found issue

    - `PARALLEL`: In this mode, TiDB will attempt to reduce the duration for collecting each batch to 1/2 of that in `DEFAULT` mode, and try to keep two TSO RPC requests working simultaneously. In this way, the average duration of each TS retrieval operation can theoretically be reduced to about 1.25 times the TSO RPC duration, which is about 83% of the time cost in `DEFAULT` mode. However, the effect of batching will be reduced, and the number of TSO RPC requests will increase to about twice that in `DEFAULT` mode.
    - `PARALLEL-FAST`: Similar to `PARALLEL` mode, in this mode, TiDB will attempt to reduce the duration for collecting each batch to 1/4 of that in `DEFAULT` mode, and try to keep 4 TSO RPC requests working simultaneously. In this way, the average duration of each TS retrieval operation can theoretically be reduced to about 1.125 times the TSO RPC duration, which is about 75% of the time cost in `DEFAULT` mode. However, the effect of batching will be further reduced, and the number of TSO RPC requests will increase to about 4 times that in `DEFAULT` mode.

> - When the following precondition is satisfied, you can consider to switch this variable to `PARALLEL` or `PARALLEL-FAST` to gain some performance improvement:

Reviewer: ```suggestion
- When the following conditions are met, you can consider switching this variable to `PARALLEL` or `PARALLEL-FAST` for potential performance improvements:
```

    - TSO waiting time constitutes a significant portion of the total execution time of SQL queries.

>     - The TSO allocation in PD doesn't reach the bottleneck.
>     - PD and TiDB nodes have sufficient CPU.
>     - The network latency between TiDB and PD is significantly higher than the time cost of PD processing TSO allocation, i.e., the network latency is the majority part of the duration of TSO RPC.

Reviewer: ```suggestion
    - The TSO allocation in PD has not reached its bottleneck.
    - PD and TiDB nodes have sufficient CPU resources.
    - The network latency between TiDB and PD is significantly higher than the time PD takes to allocate TSO (that is, network latency accounts for the majority of TSO RPC duration)
```

>         - The duration of TSO RPC requests can be observed in the **PD TSO RPC Duration** panel which can be found in the "PD Client" secion of "TiDB" Grafana dashboard; the duration of PD's TSO allocation can be observed in the **PD server TSO handle duration** panel which can be found in the "TiDB" section of "PD" Grafana dashboard.

Reviewer: ```suggestion
        - To get the duration of TSO RPC requests, check the **PD TSO RPC Duration** panel in the PD Client section of the Grafana TiDB dashboard.
        - To get the duration of PD TSO allocation, check the **PD server TSO handle duration** panel in the TiDB section of the Grafana PD dashboard.
```

>     - The additional network traffic is acceptable if the TSO RPC requests between TiDB and PD increases to twice (for `PARALLEL` mode) or 4 times (for `PARALLEL-FAST` mode) of before.

Reviewer: ```suggestion
    - The additional network traffic resulting from more TSO RPC requests between TiDB and PD (twice for `PARALLEL` or 4 times for `PARALLEL-FAST`) is acceptable.
```

> **Note:**
>

> > - `PARALLEL` mode and `PARALLEL-FAST` mode are incompatible to [`tidb_tso_client_batch_max_wait_time`](#tidb_tso_client_batch_max_wait_time-new-in-v530) 和 [`tidb_enable_tso_follower_proxy`](#tidb_enable_tso_follower_proxy-new-in-v530). If either [`tidb_tso_client_batch_max_wait_time`](#tidb_tso_client_batch_max_wait_time-new-in-v530) is set to no-zero value or [`tidb_enable_tso_follower_proxy`](#tidb_enable_tso_follower_proxy-new-in-v530) is enabled, configuring `tidb_tso_client_rpc_mode` won't take effect, and TiDB will always work in `DEFAULT` mode.

Reviewer: ```suggestion
> - `PARALLEL` and `PARALLEL-FAST` modes are incompatible with [`tidb_tso_client_batch_max_wait_time`](#tidb_tso_client_batch_max_wait_time-new-in-v530) and [`tidb_enable_tso_follower_proxy`](#tidb_enable_tso_follower_proxy-new-in-v530). If either [`tidb_tso_client_batch_max_wait_time`](#tidb_tso_client_batch_max_wait_time-new-in-v530) is set to a non-zero value or [`tidb_enable_tso_follower_proxy`](#tidb_enable_tso_follower_proxy-new-in-v530) is enabled, configuring `tidb_tso_client_rpc_mode` does take effect, and TiDB always works in `DEFAULT` mode.
```

> > - `PARALLEL` mode and `PARALLEL-FAST` mode are designed to reduce the average time cost of retrieving TS in TiDB. For situations with significant latency fluctuations, such as long-tail and latency spike issues, these two modes may not provide any remarkable performance improvement.

Reviewer: ```suggestion
> - `PARALLEL` and `PARALLEL-FAST` modes are designed to reduce the average time for retrieving TS in TiDB. In cases of situations with significant latency fluctuations, such as long-tail latency or spikes, these two modes might not provide any remarkable performance improvements.
```

> **Note:**
>
> - Suppose that the TSO RPC latency increases for reasons other than a CPU usage bottleneck of the PD leader (such as network issues). In this case, increasing the value of `tidb_tso_client_batch_max_wait_time` might increase the execution latency in TiDB and affect the QPS performance of the cluster.

> > - This feature is incompatble with [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840). If this variable is set to non-zero value, [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840) does not take effect.

Reviewer: ```suggestion
> - This feature is incompatble with [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840). If this variable is set to a non-zero value, [`tidb_tso_client_rpc_mode`](#tidb_tso_client_rpc_mode-new-in-v840) does not take effect.
```

### tidb_tso_client_rpc_mode <span class="version-mark">New in v8.4.0</span>

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Enumeration
- Default value: `DEFAULT`
- Value options: `DEFAULT`, `PARALLEL`, `PARALLEL-FAST`

> - This variable switches the mode in which TiDB sends TSO RPC requests to PD. The mode determines whether TSO RPC requests can be processed in parallel and affects the time spent on batch-waiting for each TS retrieval operation, thereby helping reduce the wait time for retrieving TS during the execution of queries in certain scenarios.

Reviewer 2: ```suggestion
- This variable switches the mode in which TiDB sends TSO RPC requests to PD. The mode determines whether TSO RPC requests are processed in parallel and affects the time spent on batch-waiting for each TS retrieval operation, thereby helping reduce the wait time for retrieving TS during query execution in certain scenarios.
```

>     - `DEFAULT`: TiDB collects TS retrieval operations during a specific period into a single TSO RPC request and sends it to PD to get timestamps in a batch. Therefore, the duration of each TS retrieval operation consists of the time spent waiting to be batched and the time spent performing the RPC. In `DEFAULT` mode, different TSO RPC requests are performed serially, and the average duration of each TS retrieval operation is about 1.5 times the actual time cost of a TSO RPC request.
>     - `PARALLEL`: In this mode, TiDB attempts to reduce the duration for collecting each batch to 1/2 of that in `DEFAULT` mode and tries to keep two TSO RPC requests running simultaneously. In this way, the average duration of each TS retrieval operation can theoretically be reduced to about 1.25 times the TSO RPC duration, which is about 83% of the time cost in `DEFAULT` mode. However, the effect of batching will be reduced, and the number of TSO RPC requests will increase to roughly double that in `DEFAULT` mode.
>     - `PARALLEL-FAST`: Similar to `PARALLEL` mode, in this mode, TiDB attempts to reduce the duration for collecting each batch to 1/4 of that in `DEFAULT` mode and tries to keep four TSO RPC requests running simultaneously. In this way, the average duration of each TS retrieval operation can theoretically be reduced to about 1.125 times the TSO RPC duration, which is about 75% of the time cost in `DEFAULT` mode. However, the effect of batching will be further reduced, and the number of TSO RPC requests will increase to roughly four times that in `DEFAULT` mode.

Reviewer 2: ```suggestion
    - `DEFAULT`: TiDB collects TS retrieval operations over a specific period into a single TSO RPC request and sends it to PD to get timestamps in a batch. Therefore, the duration of each TS retrieval operation consists of the time spent waiting to be batched and the time spent performing the RPC. In `DEFAULT` mode, different TSO RPC requests are processed serially, and the average duration of each TS retrieval operation is about 1.5 times the actual time cost of a TSO RPC request.
    - `PARALLEL`: In this mode, TiDB attempts to reduce the duration for collecting each batch to half of that in `DEFAULT` mode and tries to maintain two concurrent TSO RPC requests. In this way, the average duration of each TS retrieval operation can theoretically be reduced to about 1.25 times the TSO RPC duration, which is about 83% of the time cost in `DEFAULT` mode. However, the effect of batching will be reduced, and the number of TSO RPC requests will increase to roughly double that in `DEFAULT` mode.
    - `PARALLEL-FAST`: Similar to `PARALLEL` mode, in this mode, TiDB attempts to reduce the duration for collecting each batch to a quarter of that in `DEFAULT` mode and tries to maintain four concurrent TSO RPC requests. In this way, the average duration of each TS retrieval operation can theoretically be reduced to about 1.125 times the TSO RPC duration, which is about 75% of the time cost in `DEFAULT` mode. However, the effect of batching will be further reduced, and the number of TSO RPC requests will increase to roughly four times that in `DEFAULT` mode.
```

- When the following conditions are met, you can consider switching this variable to `PARALLEL` or `PARALLEL-FAST` for potential performance improvements:

    - TSO waiting time constitutes a significant portion of the total execution time of SQL queries.
    - The TSO allocation in PD has not reached its bottleneck.
    - PD and TiDB nodes have sufficient CPU resources.

>     - The network latency between TiDB and PD is significantly higher than the time PD takes to allocate TSO (that is, network latency accounts for the majority of TSO RPC duration)

Reviewer: ```suggestion
    - The network latency between TiDB and PD is significantly higher than the time PD takes to allocate TSO (that is, network latency accounts for the majority of TSO RPC duration).
```

        - To get the duration of TSO RPC requests, check the **PD TSO RPC Duration** panel in the PD Client section of the Grafana TiDB dashboard.
        - To get the duration of PD TSO allocation, check the **PD server TSO handle duration** panel in the TiDB section of the Grafana PD dashboard.
    - The additional network traffic resulting from more TSO RPC requests between TiDB and PD (twice for `PARALLEL` or 4 times for `PARALLEL-FAST`) is acceptable.

> **Note:**
>
> - `PARALLEL` and `PARALLEL-FAST` modes are incompatible with [`tidb_tso_client_batch_max_wait_time`](#tidb_tso_client_batch_max_wait_time-new-in-v530) and [`tidb_enable_tso_follower_proxy`](#tidb_enable_tso_follower_proxy-new-in-v530). If either [`tidb_tso_client_batch_max_wait_time`](#tidb_tso_client_batch_max_wait_time-new-in-v530) is set to a non-zero value or [`tidb_enable_tso_follower_proxy`](#tidb_enable_tso_follower_proxy-new-in-v530) is enabled, configuring `tidb_tso_client_rpc_mode` does not take effect, and TiDB always works in `DEFAULT` mode.

> > - `PARALLEL` and `PARALLEL-FAST` modes are designed to reduce the average time for retrieving TS in TiDB. In cases of situations with significant latency fluctuations, such as long-tail latency or latency spikes, these two modes might not provide any remarkable performance improvements.

Reviewer: ```suggestion
> - `PARALLEL` and `PARALLEL-FAST` modes are designed to reduce the average time for retrieving TS in TiDB. In situations with significant latency fluctuations, such as long-tail latency or latency spikes, these two modes might not provide any remarkable performance improvements.
```
